src/app/agent/page.tsx

```diff
@@ -270,133 +283,134 @@ const AgentPageComponent: React.FC = () => {
   const docs = searchParams.get("docs") !== null;
   const webrtcUrl = searchParams.get("webrtc") ?? undefined;
   const [showChooser, setShowChooser] = useState(
-    searchParams.get("chooser") !== null,
+    searchParams.get("chooser") !== null
   );
   const showInput = searchParams.get("input") !== null;
   const showOutput = searchParams.get("output") !== null;
-  const [showStats, setShowStats] = useState(
-    searchParams.get("stats") !== null,
-  );
-  const [chatManager, setChatManager] = useState<ChatManager | null>();
-  const [input, setInput] = useState("");
-  const [output, setOutput] = useState("");
-  const [helpText, setHelpText] = useState(idleText);
-  const [asrLatency, setAsrLatency] = useState(0);
-  const [llmResponseLatency, setLlmResponseLatency] = useState(0);
-  const [llmTokenLatency, setLlmTokenLatency] = useState(0);
-  const [ttsLatency, setTtsLatency] = useState(0);
-  const active = () =>
-    chatManager && chatManager!.state != ChatManagerState.IDLE;
-  useEffect(
-    () => init(),
-    [
-      asrProvider,
-      asrLanguage,
-      ttsProvider,
-      ttsModel,
-      ttsVoice,
-      model,
-      agentId,
-      docs,
-    ],
-  );
-  const init = () => {
-    console.log(
-      `[page] init asr=${asrProvider} tts=${ttsProvider} llm=${model} agent=${agentId} docs=${docs}`,
-    );
-    const manager = createChatManager({
-      asrProvider,
-      asrModel,
-      asrLanguage,
-      ttsProvider,
-      ttsModel,
-      ttsVoice,
-      model,
+
+  // Returns true if the voice session is active.
+  const active = useCallback(() => {
```

Comment:
I don't fully understand the `useCallback` usages - my understanding is that this hook is entirely a perf optimization, and as these functions are pretty minimal it's not clear why the hook is needed. Generally, unless there's some new incompatibility, I'd prefer to lift-and-shift the current code (modulo any renames) and then rework the React handling separately if needed.

Reply:
That was my original plan, but I was getting all kinds of lint errors with the existing code. The issue has to do with how these functions were called by useEffect hooks. This leads to a transitive dependency on the dependencies of the functions (since useEffect needs to declare dependencies unless you are calling the effect unconditionally on every render, which can be expensive). The useCallback allows us to memoize the callbacks unless their own dependencies change, and then, by extension, the useEffect hooks don't need to be rerun unless those callbacks have been changed.

The existing code was not doing this correctly and hence was breaking (for me, anyway).

Reply:
Here's a concrete example of what happens if I don't use `useCallback` for the `active` function (this applies to the other cases too).

* If I make the `active` function a regular function and not a callback, I get the error that this causes the `onKeyDown` callback to get recomputed on every render (since that function calls `active`):

```The 'active' function makes the dependencies of useCallback Hook (at line 438) change on every render. To fix this, wrap the definition of 'active' in its own useCallback() Hook.eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)```

* To fix this, I can have `onKeyDown` also be a function, and no longer a callback, which means it gets recomputed on every render. This causes the linter to complain that the `useEffect` hook on line 440 (`"Install our handlers, and clean them up on unmount."`) will be run on every render, since `onKeyDown` will change on every render.

* I can take out the dependency array for that `useEffect`, but that - you guessed it - causes it to be run on every render.

With the `useCallback` hooks, the `useEffect` on line 440 is only run whenever the `onKeyDown` callback changes, which would only be run if `active` (and its other dependencies) change.

The existing code was only partially declaring its dependencies, so it was not actually correct in that an effect might not run when it should have.

Basically, the general rule is this: if we're gonna have any `useEffect` hooks with functions as dependencies, those dependencies need to be callbacks themselves, with their own dependencies specified.

Reply:
OK, thanks for explaining. I'll take another look with this in mind.